storage/innobase/lock/lock0lock.cc

```diff
@@ -4253,24 +4442,38 @@ static void lock_rec_rebuild_waiting_queue(
 {
   lock_sys.assert_locked(cell);
 
-  for (lock_t *lock= first_lock; lock != NULL;
-       lock= lock_rec_get_next(heap_no, lock))
+  for (lock_t *lock= first_lock; lock != NULL;)
   {
-    if (!lock->is_waiting())
+    /* Store pointer to the next element, because if some lock is
+    bypassed, the pointer to the next lock in the current lock
+    object will be changed, as the current lock will change
+    its position in lock queue. */
+    lock_t *next= lock_rec_get_next(heap_no, lock);
+    if (!lock->is_waiting()) {
+      lock= next;
       continue;
+    }
```

Comment:
We could declare `next` in the `for` statement to simplify this a little:
```c++
for (lock_t *lock= first_lock, *next; lock; lock= next)
{
  next= lock_rec_get_next(heap_no, lock);
  if (!lock->is_waiting())
    continue;
  /* If a lock is bypassed, … */
```
This would also allow the `lock= next` statement to be omitted at the end of the body.